src/datachain/lib/listing.py

```diff
@@ -97,15 +109,68 @@ async def _scandir(client, prefix, fetch_workers) -> AsyncIterator:
     await main_task
 
 
-def list_bucket(uri: str, client_config=None, fetch_workers=FETCH_WORKERS) -> Callable:
+def list_bucket(uri: str, fetch_workers=FETCH_WORKERS, **kwargs) -> Callable:
     """
     Function that returns another generator function that yields File objects
     from bucket where each File represents one bucket entry.
     """
 
     def list_func() -> Iterator[File]:
-        config = client_config or {}
-        client, path = Client.parse_url(uri, None, **config)  # type: ignore[arg-type]
+        client, path = Client.parse_url(uri, None, **kwargs)  # type: ignore[arg-type]
         yield from iter_over_async(_scandir(client, path, fetch_workers), get_loop())
 
     return list_func
+
+
+def ls(
+    dc: "DataChain",
+    path: str,
+    recursive: Optional[bool] = True,
+    object_name="file",
+):
+    """
+    Return files by some path from DataChain instance which contains bucket listing.
+    Path can have globs.
+    If recursive is set to False, only first level children will be returned by
+    specified path
+    """
+
+    def _file_c(name: str) -> Column:
+        return Column(f"{object_name}.{name}")
+
+    dc = dc.filter(_file_c("is_latest") == true())
+
+    if recursive:
+        if not path or path == "/":
+            # root of a bucket, returning all latest files from it
+            return dc
+
+        if not glob.has_magic(os.path.basename(os.path.normpath(path))):
+            # path is not glob, so it's pointing to some directory or a specific
+            # file and we are adding proper filter for it
+            return dc.filter(
+                (_file_c("path") == path)
+                | (_file_c("path").glob(path.rstrip("/") + "/*"))
+            )
+
+        # path has glob syntax so we are returning glob filter
+        return dc.filter(_file_c("path").glob(path))
+    # returning only first level children by path
+    return dc.filter(pathfunc.parent(_file_c("path")) == path.lstrip("/").rstrip("/*"))
+
+
+def listing_dataset_name(uri: str, path: str) -> str:
+    """Returns special listing dataset name from uri and path"""
+    return f"{LISTING_PREFIX}{uri}/{path.lstrip('/')}"
+
+
+def listing_expired(created_at: datetime) -> bool:
+    """Checks if listing has expired based on it's creation date"""
+    return datetime.now(timezone.utc) > created_at + timedelta(seconds=LISTING_TTL)
+
+
+def listing_subset(ds1_name: str, ds2_name: str) -> bool:
+    """
+    Checks if one listing contains another one by comparing corresponding dataset names
+    """
+    return ds2_name.startswith(ds1_name)
```

Comment:
let's put asserts that we have both names end with `/` - otherwise this implementation is broken I think

Reply:
This will be broken if we introduce enlisting single file where listing dataset name won't end with `/`.
I will add asserts for now and this, alongside `listing_dataset_name`, will need to be changed when we introduce that.